core/geometry/src/curve/internalContexts/EllipticalArcApproximationContext.ts

```diff
@@ -0,0 +1,1092 @@
+  private _maxError: number;
+  private _originalRefinementCount: number;
+  protected constructor(ellipticalArc: Arc3d, maxError: number) {
+    super(ellipticalArc);
+    this._refinement = new OrderedSet<number>(compareFractions);
+    this._maxError = maxError > 0 ? maxError : EllipticalArcApproximationOptions.defaultMaxError;
+    this._originalRefinementCount = 0;
+  }
+  public override announceQuadrantBegin(q: QuadrantFractions): boolean {
+    this._refinement.clear();
+    for (const f of q.fractions)
+      this._refinement.add(f);
+    this._originalRefinementCount = this._refinement.size;
+    return true;
+  }
+  public updateFractionSet(childApproximation: Arc3d | LineSegment3d, f0: number, f1: number): void {
+    const childPerp = ErrorProcessor.computePrimitiveErrorXY(
+      this.ellipticalArc.clonePartialCurve(f0, f1), childApproximation,
+    );
+    if (childPerp && (childPerp.detailA.a > this._maxError))
+      this._refinement.add(Geometry.interpolate(f0, 0.5, f1));
+  }
+  public override announceQuadrantEnd(q: QuadrantFractions, _wasReversed: boolean): void {
+    q.fractions = [...this._refinement];
+  }
+  /** Whether the processor refined the sample set to decrease the approximation error. */
+  public get isRefined(): boolean {
+    return this._originalRefinementCount < this._refinement.size;
+  }
+  public getRefinedAngles(result?: number[]): number[] {
+    if (!result)
+      result = [];
+    else
+      result.length = 0;
+    for (const fraction of this._refinement) {
+      if (0.0 < fraction && fraction < 1.0) // skip start/end angle
+        result.push(this.ellipticalArc.sweep.fractionToRadians(fraction));
+    }
+    return result;
+  }
+}
+
+/**
+ * Processor for computing samples for a subdivision-based arc chain approximation.
+ * @internal
+ */
+class ArcChainSubdivisionErrorProcessor extends SubdivisionErrorProcessor {
+  public constructor(ellipticalArc: Arc3d, maxError: number) {
+    super(ellipticalArc, maxError);
+  }
+  public override announceArc(arc: Arc3d, f0: number, f1: number): void {
+    this.updateFractionSet(arc, f0, f1);
+  }
+}
+
+/**
+ * Processor for computing samples for a subdivision-based linestring approximation.
+ * @internal
+ */
+class LineStringSubdivisionErrorProcessor extends SubdivisionErrorProcessor {
+  public constructor(ellipticalArc: Arc3d, maxError: number) {
+    super(ellipticalArc, maxError);
+  }
+  public override announceChord(pt0: Point3d, pt1: Point3d, f0: number, f1: number): void {
+    this.updateFractionSet(LineSegment3d.create(pt0, pt1), f0, f1);
+  }
+}
+
+/**
+ * Intermediate class for constructing a sample-based approximation.
+ * @internal
+ */
+class ConstructionProcessor extends QuadrantFractionsProcessor {
+  protected _chain: CurveChain;
+  protected _quadrantChain?: CurveChain;
+  public constructor(ellipticalArc: Arc3d) {
+    super();
+    this._chain = ellipticalArc.sweep.isFullCircle ? Loop.create() : Path.create();
```

Comment:
so `isFullCircle` means arc sweep is from 0 to 360 and does not mean the arc is actually a circle, right?

Reply:
Correct.  `AngleSweep` is a general class for dealing with angles; it knows nothing about ellipses. `Circle` is descriptive, if perhaps a bit confusing here. Here, `isFullPeriod` would be better, however, `AngleSweep.isFullPeriod` would be incorrect in a context where the period is _not_ 2pi. So we have `AngleSweep.isFullCircle`.